torch/csrc/jit/frontend/ir_emitter.cpp

```diff
@@ -1753,22 +1794,74 @@ struct to_ir {
     };
     GatheredTypes gathered(typeParser_);
     gathered.gather(classinfo);
-    auto val = emitExpr(obj);
+    Value* val = emitExpr(obj);
     RefinementSet refinement;
-    if (gathered.types.size() == 1 &&
-        gathered.types.at(0)->isSubtypeOf(val->type()) &&
-        obj.kind() == TK_VAR) {
+
+    TORCH_CHECK(!gathered.types.empty(), "`isinstance` must be used with a "
+                "type or a tuple of types");
+
+    bool all_are_subtypes = std::all_of(gathered.types.begin(), gathered.types.end(),
+                                        [&](const TypePtr t) {
+                                          return t->isSubtypeOf(val->type());
+                                        });
+    if (all_are_subtypes && obj.kind() == TK_VAR) {
       std::string ident = Var(obj).name().name();
-      Refinement isinstance(std::move(ident), gathered.types.at(0));
-      refinement = RefinementSet({isinstance}, {});
+
+      // Get the type to compare against. If we have a tuple of types,
+      // turn the tuple into a Union type to prevent unnecessary extra
+      // logic for single vs. multiple comparisons
+      const TypePtr single_comparison = gathered.types.size() == 1 ?
+                                        gathered.types[0] :
+                                        UnionType::create(gathered.types);
+
+      Refinement isinstance(ident, single_comparison);
+
+      if (val->type()->kind() == UnionType::Kind) {
+        const UnionTypePtr union_type = val->type()->expect<UnionType>();
+
+        std::vector<TypePtr> not_isinstance_types;
+
+        // O(1) lookups of all the types in `gathered.types`
+        std::unordered_set<TypePtr, std::hash<TypePtr>, c10::TypeEqual> dict{
+          gathered.types.begin(), gathered.types.end()};
+
+        for (auto type : union_type->containedTypes()) {
+          // Fast path
+          if (!dict.count(type)) {
+            not_isinstance_types.emplace_back(type);
+            break;
+          }
+          // Check all types in `gathered.types` to see if there's a
+          // common supertype
+          for (const TypePtr comparison_type : gathered.types) {
+            c10::optional<TypePtr> unified = unifyTypes(type, comparison_type);
+            if (!unified) {
```

Comment:
Would you mind commenting with what the algorithm we're trying to do here is ? I'm having trouble following the logic.

Reply:
I'm guessing that the confusing part is only if the lhs is a `Union`, but lmk if that's not the case

For example:
```
def fn(x: Union[int, str]):
    if isinstance(x, str):
        call_str_method(x)    # x should be refined to `str` and this method should work
    …
```

This is the general algorithm:
1. Get single comparison type from rhs. (Note: we only ever have one type, but, in the next few steps, I'm going to mostly use "type" to refer to what we get from a single-element comparison, and "types" to refer to a 2+-element comparison. Of course, the latter becomes a single `Union` type.)
2. Put that comparison type in the `isinstance` vector, which says “if the lhs type is an instance of one of the rhs types, then we know we can refine it to one of the rhs types”. (We can’t know which exact rhs type the lhs type can be refined to until runtime; we can only generate code saying that it must be one of those types. A cool follow-up project would be automatically generating hidden conditional asserts to further refine the type.)
3. For each type in the lhs `Union`, see if that type does NOT share a common parent with the rhs comparison type, i.e. returns `c10::nullopt` from `unifyTypes`. This allows us to refine Unions to smaller Unions.
4. Create the correct Refinements.

Reply:
> 1. Get single comparison type from rhs. 

I'm not sure this is helpful. As in the following case:
```
def foo(x: Union[int, str]):
     if isinstance(x, (int, float)):
```
`Union[int, float]` doesn't subtype `Union[int, str]` , however we would like to retain the `isinstance` check and insert a runtime check that the input is an int with a corresponding refinement to int. 

I think it might be easier to reason about this if we decompose Union inputs into their composed types. This also unifies the type logic with the other path. 

for each type in the input, we want to see if it subtypes any of the types in `isisntance`:

* for the types that do subtype types in `isinstance`, we record which types are their supertype, and add those supertypes to the runtime checks we need to generate in the `isinstance`. the union of all the types that are subtypes will be the type refinement in the true branch.
* The union of types that do not subtype types in `isinstance` get added to the false branch type refinements

If either of the do/do not subtypes sets is exactly equal to the input then we know the isinstance check is statically true/false.


> A cool follow-up project would be automatically generating hidden conditional asserts to further refine the type 
I'm not sure what you mean here. `prim::UncheckedCast` is what refines the type in the IR if that's what you mean ?